interpreter/prune.go

```diff
@@ -76,6 +76,46 @@ func PruneAst(expr *exprpb.Expr, macroCalls map[int64]*exprpb.Expr, state EvalSt
 			maxID = id + 1
 		}
 	}
+	exprs := []*exprpb.Expr{expr}
+	for len(exprs) != 0 {
```

Comment:
Just an idea -- can we find the max from SourcePositions? Presently, I would think its keys contain all of the Expr IDs so iterating over that would be far cheaper

Reply:
If we have an optimized AST, I'm not sure we could rely on the source positions; however, we should be able to rely on updated type metadata. That said, the prune step works with both checked and parsed expressions, so I think this is the only real common denominator between each format.